### PR TITLE
threading fixes

### PR DESCRIPTION
**Pull request by sidaw**
use new Buffer, roll back extra synchronized, add resetInfos, and remove the RuntimeException on end_track

### CODE REVIEW

File: src/main/java/fig/basic/LogInfo.java [hunk 1]
```diff
@@ -49,6 +49,13 @@ private static ThreadLogInfo getInfo() {
     }
   }
 
+	// on catching exception, call this method to reset
+	// so no need to worry about IndLevel
+	public static void resetInfos() {
+		mainInfo = new ThreadLogInfo(out, fileOut);
+		threadInfos = new ArrayList<ThreadLogInfo>();
```
Comment (percyliang) [inline src/main/java/fig/basic/LogInfo.java]: leave null if already null

File: src/main/java/fig/basic/LogInfo.java [hunk 2]
```diff
@@ -235,7 +242,8 @@ public synchronized void begin_track_general(Object o,
 
   public synchronized void end_track() {
     if (indLevel == 0) {
-      throw new RuntimeException("Already at indLevel = 0, can't decrement (you probably have too many end_track's)");
+			indLevel = 1;
```
Comment (percyliang) [inline src/main/java/fig/basic/LogInfo.java]: indentation

**Comment by sidaw** [comment]
okay, replaced tab with 2 spaces, including 2 not due to me.

File: src/main/java/fig/basic/LogInfo.java [hunk 3]
```diff
@@ -49,6 +49,14 @@ private static ThreadLogInfo getInfo() {
     }
   }
 
+  // on catching exception, call this method to reset
+  // so no need to worry about IndLevel
```
Comment (percyliang) [inline src/main/java/fig/basic/LogInfo.java]: nit: use consistent case: `indLevel`

**Comment by sidaw** [comment]
Okay. Still have to actually merge this PR, instead of just approve.

**Pull request merged**